### DISCUSSION TITLE
Clickhouse write performance degrades over time

### DISCUSSION THREAD

**Discussion by ptacek** [ClickHouse/ClickHouse#36419] [Q&A] [answered]
I'm using built-in Kafka engine with materialized views to sink messages from Kafka to Clickhouse tables (Replicated merge tree). I'm really amazed how many messages per second can Clickhouse handle, but the problem in my environment is, that it doesn't last for a long time.

For a first few days, everything is perfectly OK and then, after approximately 10 days, Kafka consumer group's lag (difference between max offset and last committed offset) starts growing for the biggest topics. The lag is then getting worse every day. My only currently working "solution" is to restart Clickhouse service on each server. Lag of all consumer groups begins to decrease immediately and everything is fine again for a few days.

Consumer lag of the bigest topic:
![JK9ij](https://user-images.githubusercontent.com/5675755/163994673-61329c48-5040-4b56-b7b8-9d9a35660c4d.png)

Clickhouse version: 22.3.2.1, there are ~500K messages per second in Kafka cluster. One of the biggest topics has ~100K messages per second (in 32 partitions). Number of messages is lower during night hours. That's why the lag is decreasing during that time. Kafka cluster has 4 servers. CH cluster has 12 servers. Zookeeper is on 3 of the CH servers, but at least it has its own dedicated SSDs. I'm also utilizing Clickhouse feature for multiple storages, so that the most recent data are stored on SSDs and cold data are being moved to HDDs.

Do you have any idea why might this be happening? What settings or metrics to check? Restarting Clickhouse every second week is not a sustainable solution.

One of the things that i have noticed are timeouts in error log, that starts appearing over time, but i have no idea why:
`<Warning> StorageKafka (----): [rdk:SESSTMOUT] [thrd:main]: Consumer group session timed out (in join-state steady) after 10402 ms without a successful response from the group coordinator (broker 140, last error was Broker: Group rebalance in progress): revoking assignment and rejoining group`

And also chart with memory consumption, that is constantly increasing until restarts:
![Htvbb](https://user-images.githubusercontent.com/5675755/163994845-4e51ee9f-fec0-4fab-ab96-c28822c76407.png)

**Comment by seanlaff** [comment]
Are you writing to a distributed table? Possibly unrelated but when I was slamming a distributed table there'd be some funny io/fsync business at some point. I swapped to `insert_distributed_sync = 1` to fix my issue.

**Comment by ptacek** [comment]
I'm trying both writing to distributed table and to local table. Depending on the topic/table

**Comment by ptacek** [comment]
So I tried to write only to local tables instead of distributed tables. Also tried to increase max_poll_interval_ms and session_timeout_ms and restarted the cluster. But after 11 days, the Kafka consumers started lagging again and the overall performance has decreased.

Now I'm upgrading to 22.4.4.7, hoping that this bugfix https://github.com/ClickHouse/ClickHouse/pull/35973 and increasing number of consumers might help. We'll see in 2 weeks

**Reply by UnamedRus** [reply]
How many partiitons you have in kafka, how many Kafka tables reading this topic you have and how many consumers per kafka table

**Reply by ptacek** [reply]
There are 4 bigger topics (~100K messages/s) that have 36 partitions each. Every topic has its corresponding table on every of 12 servers in the cluster.  Before the last CH upgrade, every kafka table had kafka_num_consumers = 1. After the upgrade i've increased it to 3, so i hope that it makes 12 x 3 = 36 consumers - one for each topic partition.

Then there are about 30 another smaller topics with 4-12 partitions each and 100K messages/s in total. Their all kafka tables have kafka_num_consumers = 1.

**Comment by ptacek** [comment]
So upgrading to version with the bugfix https://github.com/ClickHouse/ClickHouse/pull/35973 and increasing number of kafka consumers didn't help. After 11 days after Clickhouse restart, the insert performance started decreasing again and consumer lag is raising. 

What is interesting is that the lag is significantly different for consumers running on different servers. But the servers in the cluster are all the same, running just instances of Clickhouse. Data are distributed into partitions equally and 1 consumer = 1 partition.
For example consumers running on one server have already lag above 100M offsets. Consumers running on another Clickhouse server have lag around 30-40M offsets and some consumers, running on different servers are not lagging at all.

Tomorrow i'm going to upgrade CH to version 22.6. Restart of the cluster will cause the lag to drop again, but I'm sure, that it will start happening after 10-11 days again. 
Do you have any suggestions what else to try? 
I'm thinking about trying to stop using different volumes for hot/cold data in one table and see if it makes any difference in this problem.
Might migration to Clickhouse Keeper be helpful?

**Reply by MANnsz** [reply]
I have the same problem. Did you solve it after upgrade ch version to 22.6?

**Reply by ptacek** [reply]
Unfortunatelly not. I still have to restart the cluster every 2 weeks

**Comment by ptacek** [comment]
14 days ago, I upgraded Clickhouse to version 22.7.  and kafka consumer lags are still low. It has never lasted in good condition for so long before. **So far, it looks, that there was something in 22.7. release, that fixed my problem**. I will keep monitoring it

**Comment by ptacek** [comment, marked as answer]
Ok, so it's been 2 months since the upgrade to version 22.7. Never restarted or upgraded again since then and everything works perfectly. I consider this as a solution to my problem.

**Comment by journaux** [comment]
thanks for the story - have you tried experimenting @ even larger stress levels @ptacek? 
we want to make sure a reasonably small, partitioned cluster can consume 50mm iops / sec
w/out major degradation

**Comment by karanbalani** [comment]
Hey, @ptacek 

Please share a sample of your setup. It could help others following this thread or encountering something similar. It could be a good starting for many people because you have handled a good scale for this. ✌️
